### PR TITLE
feat(elasticache): add ElastiCache Serverless nuking

### PR DESCRIPTION
**Pull request by Petersoj**

### CODE REVIEW

**Comment by Petersoj** [comment]
Will fix to pass tests tomorrow morning.

**Comment by Petersoj** [comment]
I removed tag listings for serverless caches since AWS throws when attempting to query the tags of a serverless cache that doesn't currently have the "available" status.

**Comment by ekristen** [comment]
I'm wondering if this should be its own resource. Looking at it on mobile, will look from my laptop soon.

**Comment by Petersoj** [comment]
I suppose it could be, but I think of serverless and provisioned/on-demand clusters as essentially the same resource in this context. Similar to how RDS also has serverless and provisioned/on-demand instances in a cluster.

Also, thoughts on removing the "memcache" name from the file since ElastiCache supports both Memcached and Redis OSS and the SDK calls don't differentiate between the cache type? Just a minor detail.

File: resources/elasticache-memcacheclusters.go [hunk 1]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
 		}
 
 		resources = append(resources, &ElasticacheCacheCluster{
-			svc:       svc,
-			clusterID: cacheCluster.CacheClusterId,
-			status:    cacheCluster.CacheClusterStatus,
-			Tags:      tags.TagList,
+			svc:        svc,
+			clusterID:  cacheCluster.CacheClusterId,
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:66]:
```suggestion
			ClusterID:  cacheCluster.CacheClusterId,
```

Reason: libnuke now has this automatic properties generator based on the resource struct.

File: resources/elasticache-memcacheclusters.go [hunk 2]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+			status:     cacheCluster.CacheClusterStatus,
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:67]:
```suggestion
			Status:     cacheCluster.CacheClusterStatus,
```

Reason: libnuke now has this automatic properties generator based on the resource struct.

File: resources/elasticache-memcacheclusters.go [hunk 3]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+			Tags:       tags.TagList,
+			serverless: false,
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:69]: Just remove this line here.

Reason: by default it is false.

File: resources/elasticache-memcacheclusters.go [hunk 4]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
 	return resources, nil
 }
 
 type ElasticacheCacheCluster struct {
-	svc       elasticacheiface.ElastiCacheAPI
-	clusterID *string
-	status    *string
-	Tags      []*elasticache.Tag
+	svc        elasticacheiface.ElastiCacheAPI
+	clusterID  *string
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:93]:
```suggestion
	ClusterID  *string
```

File: resources/elasticache-memcacheclusters.go [hunk 5]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+	status     *string
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:94]:
```suggestion
	Status     *string
```

File: resources/elasticache-memcacheclusters.go [hunk 6]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+	Tags       []*elasticache.Tag
+	serverless bool
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:96]:
```suggestion
	Serverless bool
```

File: resources/elasticache-memcacheclusters.go [hunk 7]
```diff
@@ -85,22 +102,36 @@ func (i *ElasticacheCacheCluster) Properties() types.Properties {
...
 
-	for _, tag := range i.Tags {
-		properties.SetTag(tag.Key, tag.Value)
+	if i.Tags != nil {
+		for _, tag := range i.Tags {
+			properties.SetTag(tag.Key, tag.Value)
+		}
 	}
 
 	return properties
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:111]:
```suggestion
	return types.NewPropertiesFromStruct(i)
```

TODO: remove everything else from the Properties() method.

Reason: libnuke now has this automatic properties generator based on the resource struct.

File: resources/elasticache-memcacheclusters.go [hunk 8]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
-	status    *string
-	Tags      []*elasticache.Tag
+	svc        elasticacheiface.ElastiCacheAPI
+	clusterID  *string
+	status     *string
+	Tags       []*elasticache.Tag
+	serverless bool
 }
 
 func (i *ElasticacheCacheCluster) Properties() types.Properties {
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:99]:
```suggestion
func (r *ElasticacheCacheCluster) Properties() types.Properties {
```

Reason: standardizing all pointers for resources to `r`

File: resources/elasticache-memcacheclusters.go [hunk 9]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+	serverlessParams := &elasticache.DescribeServerlessCachesInput{MaxResults: aws.Int64(100)}
+	serverlessResp, serverlessErr := svc.DescribeServerlessCaches(serverlessParams)
+	if serverlessErr != nil {
+		return nil, serverlessErr
+	}
+
+	for _, serverlessCache := range serverlessResp.ServerlessCaches {
+		resources = append(resources, &ElasticacheCacheCluster{
+			svc:        svc,
+			clusterID:  serverlessCache.ServerlessCacheName,
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:82]:
```suggestion
			ClusterID:  serverlessCache.ServerlessCacheName,
```

File: resources/elasticache-memcacheclusters.go [hunk 10]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+			status:     serverlessCache.Status,
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:83]:
```suggestion
			Status:     serverlessCache.Status,
```

File: resources/elasticache-memcacheclusters.go [hunk 11]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+			serverless: true,
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:84]:
```suggestion
			Serverless: true,
```

File: resources/elasticache-memcacheclusters.go [hunk 12]
```diff
@@ -85,22 +102,36 @@ func (i *ElasticacheCacheCluster) Properties() types.Properties {
...
+	if i.Tags != nil {
+		for _, tag := range i.Tags {
+			properties.SetTag(tag.Key, tag.Value)
+		}
 	}
 
 	return properties
 }
 
 func (i *ElasticacheCacheCluster) Remove(_ context.Context) error {
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:114]: If would change this to the following:

if serverless return DeleteServerlessCacheInput

drop the else 

and then run the deleteCacheCluster

A bit cleaner this way.

File: resources/elasticache-memcacheclusters_mock_test.go [hunk 13]
```diff
@@ -26,8 +26,9 @@ func Test_Mock_ElastiCache_CacheCluster_Remove(t *testing.T) {
 	mockElastiCache := mock_elasticacheiface.NewMockElastiCacheAPI(ctrl)
 
 	cacheCluster := ElasticacheCacheCluster{
-		svc:       mockElastiCache,
-		clusterID: aws.String("foobar"),
+		svc:        mockElastiCache,
+		clusterID:  aws.String("foobar"),
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters_mock_test.go:30]:
```suggestion
		ClusterID:  aws.String("foobar"),
```

File: resources/elasticache-memcacheclusters_mock_test.go [hunk 14]
```diff
@@ -26,8 +26,9 @@ func Test_Mock_ElastiCache_CacheCluster_Remove(t *testing.T) {
...
+		serverless: false,
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters_mock_test.go:31]:
```suggestion
		Serverless: false,
```

File: resources/elasticache-memcacheclusters.go [hunk 15]
```diff
@@ -62,21 +62,38 @@ func (l *ElasticacheCacheClusterLister) List(_ context.Context, o interface{}) (
...
+		})
+	}
+
+	serverlessParams := &elasticache.DescribeServerlessCachesInput{MaxResults: aws.Int64(100)}
+	serverlessResp, serverlessErr := svc.DescribeServerlessCaches(serverlessParams)
+	if serverlessErr != nil {
+		return nil, serverlessErr
+	}
+
+	for _, serverlessCache := range serverlessResp.ServerlessCaches {
```
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:79]: You need to list tags for serverless too.

File: resources/elasticache-memcacheclusters.go (hunk 12 above)
Comment (ekristen) [inline resources/elasticache-memcacheclusters.go:114]:
```suggestion
func (r *ElasticacheCacheCluster) Remove(_ context.Context) error {
```

**Review by ekristen — CHANGES_REQUESTED** [review]
@Petersoj I reviewed this based on everything I would do if I were updating this or touching this to make updates. 

I agree it should stay the same resource after looking at the UI/API and how they are distinguished. There is an argument for making them separate but for now I think they can remain the same.

If you'd prefer I can just do a commit over top of your changes already, or if you'd like to take a stab. 

I did everything in suggestion format to make it a bit easier on you. Let me know.

**Comment by Petersoj** [comment]
@ekristen Thank you for the comprehensive review! Feel free to commit those changes in my PR, along with any other changes you see fit. I agree with all your comments. I have pretty much zero experience in Go haha so I appreciate you taking time to give me an in-depth review.

Regarding the tags issue, I observed that AWS throws when querying the tags after initiating the deletion process. Querying the tags during the "available" state is no problem. So putting in a guard to check for the "available" state when querying the tags seems appropriate as this would still allow for users to filter based on those tags during the initial listings iirc. According to the [docs](https://docs.aws.amazon.com/AmazonElastiCache/latest/APIReference/API_ServerlessCache.html), there are the following states:
> CREATING, AVAILABLE, DELETING, CREATE-FAILED and MODIFYING

**Comment by Petersoj** [comment]
Did some manual testing and discovered the following API behavior of AWS when querying tags of a serverless ElastiCache cluster in various states:
| Status | `listTagsForResource` call result (via Java SDK) |
|--------|--------|
| `CREATING` | InvalidServerlessCacheStateException: Serverless cache is not in an available state to perform the requested action. (Service: ElastiCache, Status Code: 400, Request ID: a0f227b9-f956-44ae-b5de-e08be3233b87) |
| `AVAILABLE` | Success |
| `DELETING` | InvalidServerlessCacheStateException: Serverless cache is not in an available state to perform the requested action. (Service: ElastiCache, Status Code: 400, Request ID: 51c8a879-5c4f-4e77-a39c-b22b107ccb85) |
| `CREATE-FAILED` (not sure how to trigger; perhaps an invalid cluster parameter?) | I suspect AWS will throw |
| `MODIFYING` (triggered via changing access control group) | Success |

The AWS console doesn't display an error on the "Tags" tab of a serverless cluster, but the network request it makes results in a 400 status code.
console
So `AVAILABLE` and `MODIFYING` appear to be the only valid statuses the tags can be queried in.
Odd behavior from AWS...

**Comment by Petersoj** [comment]
@ekristen thoughts?

**Comment by ekristen** [comment]
@Petersoj this is great, thank you. I've been pretty swamped this week, looking to come back around to this this weekend.

**Comment by Petersoj** [comment]
Ok awesome. No rush. Thanks!

**Comment by ekristen** [comment]
@Petersoj I've pushed my changes. Please take a look and test it out, let me know if you run into problems.

**Comment by Petersoj** [comment]
Just tested it and it works great. Thanks!

**Review by ekristen — APPROVED** [review]

**Pull request merged**